File: backend/app/api/endpoints/schedule_assignments.py
from fastapi import APIRouter, Depends, HTTPException, status
from typing import List
from app.api.deps import get_supabase_client
from app.repositories.schedule_assignments_repository import ScheduleAssignmentsRepository
from app.services.schedule_assignments_service import ScheduleAssignmentsService
from app.schemas.schedule_assignments import (
    ScheduleAssignmentCreate,
    ScheduleAssignmentUpdate,
    ScheduleAssignmentResponse,
    ScheduleAssignmentWithDetails
)
from supabase import Client

router = APIRouter()


@router.get("/", response_model=List[ScheduleAssignmentResponse])
async def get_all_assignments(client: Client = Depends(get_supabase_client)):
    """すべてのスケジュール割り当てを取得"""
    try:
        service = ScheduleAssignmentsService(ScheduleAssignmentsRepository(client))
        assignments = await service.get_all_assignments()
        return assignments
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/practice-slot/{practice_slot_id}", response_model=List[ScheduleAssignmentWithDetails])
async def get_assignments_by_practice_slot(
    practice_slot_id: str, 
    client: Client = Depends(get_supabase_client)
):
    """指定された練習日のすべての割り当てを取得"""
    try:
        service = ScheduleAssignmentsService(ScheduleAssignmentsRepository(client))
        assignments = await service.get_assignments_by_practice_slot(practice_slot_id)
        return assignments
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/practice-slot/{practice_slot_id}/time/{time_slot}/group/{group_id}", response_model=ScheduleAssignmentWithDetails)
async def get_assignment_by_time_and_group(
    practice_slot_id: str,
    time_slot: str,
    group_id: str,
    client: Client = Depends(get_supabase_client)
):
    """指定された時間・グループの割り当てを取得"""
    try:
        service = ScheduleAssignmentsService(ScheduleAssignmentsRepository(client))
        assignment = await service.get_assignment_by_time_and_group(practice_slot_id, time_slot, group_id)
        if not assignment:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Assignment not found")
        return assignment
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/", response_model=ScheduleAssignmentResponse)
async def create_assignment(
    assignment_data: ScheduleAssignmentCreate,
    client: Client = Depends(get_supabase_client)
):
    """新しいスケジュール割り当てを作成"""
    try:
        # デバッグ用ログ: 受信したデータを確認
        print(f"Received assignment data: {assignment_data}")
        
        service = ScheduleAssignmentsService(ScheduleAssignmentsRepository(client))
        assignment = await service.create_assignment(assignment_data)
        return assignment
    except ValueError as e:
        # バリデーションエラーの場合は422を返す
        print(f"Validation error creating assignment: {e}")
        raise HTTPException(status_code=422, detail=f"Validation error: {str(e)}")
    except Exception as e:
        print(f"Error creating assignment: {e}")
        print(f"Error type: {type(e)}")
        import traceback
        traceback.print_exc()
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")


@router.put("/{assignment_id}", response_model=ScheduleAssignmentResponse)
async def update_assignment(
    assignment_id: str,
    assignment_data: ScheduleAssignmentUpdate,
    client: Client = Depends(get_supabase_client)
):
    """スケジュール割り当てを更新"""
    try:
        service = ScheduleAssignmentsService(ScheduleAssignmentsRepository(client))
        assignment = await service.update_assignment(assignment_id, assignment_data)
        return assignment
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/{assignment_id}")
async def delete_assignment(
    assignment_id: str,
    client: Client = Depends(get_supabase_client)
):
    """スケジュール割り当てを削除"""
    try:
        service = ScheduleAssignmentsService(ScheduleAssignmentsRepository(client))
        success = await service.delete_assignment(assignment_id)
        if not success:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Assignment not found")
        return {"message": "Assignment deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/upsert", response_model=ScheduleAssignmentResponse)
async def upsert_assignment(
    assignment_data: ScheduleAssignmentCreate,
    client: Client = Depends(get_supabase_client)
):
    """スケジュール割り当てを更新または作成"""
    try:
        service = ScheduleAssignmentsService(ScheduleAssignmentsRepository(client))
        assignment = await service.upsert_assignment(assignment_data)
        return assignment
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/lib/api/schedule-assignments.ts
import { 
  ScheduleAssignment, 
  ScheduleAssignmentCreate, 
  ScheduleAssignmentUpdate, 
  ScheduleAssignmentWithDetails,
  ApiResponse 
} from '@/features/practice-slots/types/schedule-assignments';

const API_BASE_URL = 'http://localhost:8000/api/v1';

export class ScheduleAssignmentsAPI {
  private baseUrl: string;

  constructor(baseUrl: string = API_BASE_URL) {
    this.baseUrl = baseUrl;
  }

  async getAllAssignments(): Promise<ApiResponse<ScheduleAssignment[]>> {
    try {
      const response = await fetch(`${this.baseUrl}/schedule-assignments/`);
      
      if (!response.ok) {
        console.error('Schedule Assignments API error:', response.status, response.statusText);
        return {
          success: false,
          error: `API Error: ${response.status} ${response.statusText}`
        };
      }
      
      const data = await response.json();
      if (Array.isArray(data)) {
        return { success: true, data: data };
      }
      return data;
    } catch (error) {
      console.error('Error fetching schedule assignments:', error);
      return {
        success: false,
        error: 'Failed to fetch schedule assignments - Network error'
      };
    }
  }

  async getAssignmentsByPracticeSlot(practiceSlotId: string): Promise<ApiResponse<ScheduleAssignmentWithDetails[]>> {
    try {
      const response = await fetch(`${this.baseUrl}/schedule-assignments/practice-slot/${practiceSlotId}`);
      
      if (!response.ok) {
        console.error('Schedule Assignments API error:', response.status, response.statusText);
        return {
          success: false,
          error: `API Error: ${response.status} ${response.statusText}`
        };
      }
      
      const data = await response.json();
      if (Array.isArray(data)) {
        return { success: true, data: data };
      }
      return data;
    } catch (error) {
      console.error('Error fetching schedule assignments by practice slot:', error);
      return {
        success: false,
        error: 'Failed to fetch schedule assignments - Network error'
      };
    }
  }

  async getAssignmentByTimeAndGroup(
    practiceSlotId: string, 
    timeSlot: string, 
    groupId: string
  ): Promise<ApiResponse<ScheduleAssignmentWithDetails>> {
    try {
      const response = await fetch(
        `${this.baseUrl}/schedule-assignments/practice-slot/${practiceSlotId}/time/${timeSlot}/group/${groupId}`
      );
      
      if (!response.ok) {
        if (response.status === 404) {
          return { success: false, error: 'Assignment not found' };
        }
        console.error('Schedule Assignments API error:', response.status, response.statusText);
        return {
          success: false,
          error: `API Error: ${response.status} ${response.statusText}`
        };
      }
      
      const data = await response.json();
      return { success: true, data: data };
    } catch (error) {
      console.error('Error fetching schedule assignment:', error);
      return {
        success: false,
        error: 'Failed to fetch schedule assignment - Network error'
      };
    }
  }

  async createAssignment(assignmentData: ScheduleAssignmentCreate): Promise<ApiResponse<ScheduleAssignment>> {
    try {
      const response = await fetch(`${this.baseUrl}/schedule-assignments/`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(assignmentData),
      });
      
      if (!response.ok) {
        console.error('Schedule Assignments API error:', response.status, response.statusText);
        return {
          success: false,
          error: `API Error: ${response.status} ${response.statusText}`
        };
      }
      
      const data = await response.json();
      return { success: true, data: data };
    } catch (error) {
      console.error('Error creating schedule assignment:', error);
      return {
        success: false,
        error: 'Failed to create schedule assignment - Network error'
      };
    }
  }

  async updateAssignment(assignmentId: string, assignmentData: ScheduleAssignmentUpdate): Promise<ApiResponse<ScheduleAssignment>> {
    try {
      const response = await fetch(`${this.baseUrl}/schedule-assignments/${assignmentId}`, {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(assignmentData),
      });
      
      if (!response.ok) {
        console.error('Schedule Assignments API error:', response.status, response.statusText);
        return {
          success: false,
          error: `API Error: ${response.status} ${response.statusText}`
        };
      }
      
      const data = await response.json();
      return { success: true, data: data };
    } catch (error) {
      console.error('Error updating schedule assignment:', error);
      return {
        success: false,
        error: 'Failed to update schedule assignment - Network error'
      };
    }
  }

  async deleteAssignment(assignmentId: string): Promise<ApiResponse<void>> {
    try {
      const response = await fetch(`${this.baseUrl}/schedule-assignments/${assignmentId}`, {
        method: 'DELETE',
      });
      
      if (!response.ok) {
        console.error('Schedule Assignments API error:', response.status, response.statusText);
        return {
          success: false,
          error: `API Error: ${response.status} ${response.statusText}`
        };
      }
      
      return { success: true };
    } catch (error) {
      console.error('Error deleting schedule assignment:', error);
      return {
        success: false,
        error: 'Failed to delete schedule assignment - Network error'
      };
    }
  }

  async upsertAssignment(assignmentData: ScheduleAssignmentCreate): Promise<ApiResponse<ScheduleAssignment>> {
    try {
      const response = await fetch(`${this.baseUrl}/schedule-assignments/upsert`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(assignmentData),
      });
      
      if (!response.ok) {
        console.error('Schedule Assignments API error:', response.status, response.statusText);
        return {
          success: false,
          error: `API Error: ${response.status} ${response.statusText}`
        };
      }
      
      const data = await response.json();
      return { success: true, data: data };
    } catch (error) {
      console.error('Error upserting schedule assignment:', error);
      return {
        success: false,
        error: 'Failed to upsert schedule assignment - Network error'
      };
    }
  }
}

export const scheduleAssignmentsAPI = new ScheduleAssignmentsAPI();




